misk/src/test/kotlin/misk/cloud/gcp/testing/FakeHttpTransport.kt

```diff
@@ -0,0 +1,85 @@
+package misk.cloud.gcp.testing
+
+import com.google.api.client.http.HttpTransport
+import com.google.api.client.http.LowLevelHttpRequest
+import com.google.api.client.http.LowLevelHttpResponse
+import com.google.api.client.json.jackson2.JacksonFactory
+import com.google.api.client.testing.http.MockLowLevelHttpResponse
+import java.io.ByteArrayInputStream
+import java.io.ByteArrayOutputStream
+
+typealias FakeHttpResponse = MockLowLevelHttpResponse
+typealias FakeHttpResponseFactory = (FakeHttpRequest) -> FakeHttpResponse
+typealias HttpRequestMatcher = (FakeHttpRequest) -> Boolean
+typealias HttpRequestResponse = Pair<HttpRequestMatcher, FakeHttpResponseFactory>
+
+class FakeHttpTransport internal constructor(
+        val requestResponses: List<HttpRequestResponse>
+) : HttpTransport() {
+    class Builder {
+        private val requestResponses: MutableList<HttpRequestResponse> = ArrayList()
+
+        fun onRequest(
+                matcher: HttpRequestMatcher,
+                response: FakeHttpResponseFactory
+        ): FakeHttpTransport.Builder {
+            requestResponses.add(matcher to response)
+            return this
+        }
+
+        fun onRequest(matcher: HttpRequestMatcher, response: FakeHttpResponse) =
+                onRequest(matcher, { response })
+
+        fun onRequest(matcher: HttpRequestMatcher, statusCode: Int, content: Any) =
+                onRequest(matcher, FakeHttpResponse()
+                        .setStatusCode(statusCode)
+                        .setContent(JacksonFactory().toByteArray(content)))
+
+        fun onRequest(matcher: HttpRequestMatcher, statusCode: Int) =
+                onRequest(matcher, FakeHttpResponse().setStatusCode(statusCode))
+
+        fun build(): FakeHttpTransport = FakeHttpTransport(requestResponses.toList())
+
+    }
+
+    override fun buildRequest(method: String, url: String): LowLevelHttpRequest =
+            FakeHttpRequest(method, url, requestResponses)
+}
+
+class FakeHttpRequest internal constructor(
+        val method: String,
+        val url: String,
+        private val requestResponses: List<HttpRequestResponse>
+) : LowLevelHttpRequest() {
+    private val headers = HashMap<String, String>()
```

Comment:
nit: LinkedHashMap. For life

Reply:
or `mapOf()`

Reply:
can't be mapOf, has to be mutable. LinkedHashMap it is.